web/src/stream_color.js

```diff
@@ -1,41 +1,69 @@
-    const style = color;
+extend([lchPlugin, mixPlugin]);
 
-    const $stream_labels = table.find(".stream_label");
+export function update_stream_recipient_color($stream_header) {
+    if ($stream_header.length) {
+        const stream_id = Number.parseInt($($stream_header).attr("data-stream-id"), 10);
+        const stream_name = stream_data.maybe_get_stream_name(stream_id);
+        if (!stream_name) {
+            return;
+        }
+        const stream_color = stream_data.get_color(stream_name);
+        const recipient_bar_color = get_recipient_bar_color(stream_color);
+        $stream_header
+            .find(".message-header-contents")
+            .css("background-color", recipient_bar_color);
+    }
+}
+
+export function get_stream_privacy_icon_color(color) {
+    // LCH stands for Lightness, Chroma, and Hue.
+    // This function restricts Lightness of a color to be between 20 to 75.
+    color = colord(color).toLch();
+    const min_color_l = 20;
+    const max_color_l = 75;
+    if (color.l < min_color_l) {
+        color.l = min_color_l;
+    } else if (color.l > max_color_l) {
+        color.l = max_color_l;
+    }
+    return colord(color).toHex();
+}
+
+export function get_recipient_bar_color(color) {
+    // Mixes 20% of color to 80% of white (light theme) / black (dark theme).
+    const using_dark_theme = settings_data.using_dark_theme();
+    color = get_stream_privacy_icon_color(color);
+    return colord(using_dark_theme ? "#000000" : "#ffffff")
+        .mix(color, using_dark_theme ? 0.5 : 0.4)
+        .toHex();
+}
 
+function update_table_stream_color(table, stream_name, color) {
+    const recipient_bar_color = get_recipient_bar_color(color);
+    const $stream_labels = table.find(".stream_label");
     for (const label of $stream_labels) {
         const $label = $(label);
         if ($label.text().trim() === stream_name) {
-            const $messages = $label.closest(".recipient_row").children(".message_row");
-            $messages
-                .children(".messagebox")
-                .css(
-                    "box-shadow",
-                    "inset 2px 0px 0px 0px " + style + ", -1px 0px 0px 0px " + style,
-                );
-            $messages
-                .children(".date_row")
-                .css(
-                    "box-shadow",
-                    "inset 2px 0px 0px 0px " + style + ", -1px 0px 0px 0px " + style,
-                );
-            $label.css({background: style, "border-left-color": style});
-            $label.removeClass("dark_background");
-            $label.addClass(color_class.get_css_class(color));
+            $label
+                .closest(".message_header_stream .message-header-contents")
+                .css({background: recipient_bar_color});
         }
     }
 }
 
 function update_stream_privacy_color(id, color) {
+    color = get_stream_privacy_icon_color(color);
     $(`.stream-privacy-${CSS.escape(id)}`).css("color", color);
```

Comment:
This class is already used in five places. Here you’ve changed the color that they’ll update to when stream colors changed, but you haven’t changed the color they’ll initialize to on page load.

```console
$ git grep 'stream-privacy-{{' | cat
web/templates/recent_topic_row.hbs:                <span class="stream-privacy-{{stream_id}} stream-privacy filter-icon" style="color: {{stream_color}}">
web/templates/settings/stream_specific_notification_row.hbs:        <span class="stream-privacy-{{stream.stream_id}} stream-privacy filter-icon" style="color: {{stream.color}}">
web/templates/stream_sidebar_actions.hbs:            <span class="stream-privacy-{{stream.stream_id}} stream-privacy filter-icon" style="color: {{stream.color}}">
web/templates/stream_sidebar_row.hbs:            <span class="stream-privacy-{{id}} stream-privacy filter-icon" style="color: {{color}}">
web/templates/user_stream_list_item.hbs:        <span class="stream-privacy-{{stream_id}} stream-privacy filter-icon" style="color: {{stream_color}}">
```

Reply:
yeah, thanks! Will fix.